wire/decoder.go

```diff
@@ -0,0 +1,367 @@
+	"time"
+)
+
+//integer sizes
+const (
+	ByteSize    = 1
+	ShortSize   = 2
+	IntegerSize = 4
+	LongSize    = 8
+)
+
+// ConnInfo contains information about the database connection. This is returned
+// from the database on a successful login.
+type ConnInfo struct {
+
+	// Host is the host ID of the volt node
+	Host int32
+
+	// Connection unique id for the connection in the current volt node
+	Connection int64
+
+	// Time  in which the cluster was started
+	ClusterStart time.Time
+
+	// IPV4 address of the leader node
+	LeaderAddr struct {
+		Value int32
+		IP    net.IP
+	}
+
+	// Representation of the build of the connected node
+	Build string
+}
+
+// Decoder is voltdb wire protocol decoder
+type Decoder struct {
+	r io.Reader
+}
+
+// NewDecoder returns a new Decoder that decods values read from src
+func NewDecoder(src io.Reader) *Decoder {
+	return &Decoder{r: src}
+}
+
+// SetReader replaces the underlying reader, next call to Decode methods will
+// read from this
+func (d *Decoder) SetReader(r io.Reader) {
+	d.r = r
+}
+
+func (d *Decoder) Reset() {
+	d.r = nil
+}
+
+func (d *Decoder) Int32() (int32, error) {
+	u, err := d.Uint32()
+	if err != nil {
+		return 0, err
+	}
+	return int32(u), nil
+}
+
+func (d *Decoder) Uint32() (uint32, error) {
+	b, err := d.read(IntegerSize)
+	if err != nil {
+		return 0, err
+	}
+	return endian.Uint32(b), nil
+}
+
+func (d *Decoder) Int64() (int64, error) {
+	u, err := d.Uint64()
+	if err != nil {
+		return 0, err
+	}
+	return int64(u), nil
+}
+
+func (d *Decoder) Uint64() (uint64, error) {
+	b, err := d.read(LongSize)
+	if err != nil {
+		return 0, err
+	}
+	return endian.Uint64(b), nil
+}
+
+func (d *Decoder) Time() (time.Time, error) {
+	v, err := d.Int64()
+	if err != nil {
+		return time.Time{}, err
+	}
+	if v != math.MinInt64 {
+		ts := time.Unix(0, v*int64(time.Microsecond))
+		return ts.Round(time.Microsecond), err
+	}
+	return time.Time{}, nil
+}
+
+func (d *Decoder) Float64() (float64, error) {
+	v, err := d.Uint64()
+	if err != nil {
+		return 0, err
+	}
+	return math.Float64frombits(v), nil
+}
+
+func (d *Decoder) String() (string, error) {
+	length, err := d.Int32()
+	if err != nil {
+		return "", err
+	}
+	if length == -1 {
+		return "", nil
+	}
+	b := make([]byte, length)
+	_, err = d.r.Read(b)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func (d *Decoder) Uint16() (uint16, error) {
+	b, err := d.read(ShortSize)
+	if err != nil {
+		return 0, err
+	}
+	return endian.Uint16(b), nil
+}
+
+func (d *Decoder) Int16() (int16, error) {
+	v, err := d.Uint16()
+	if err != nil {
+		return 0, err
+	}
+	return int16(v), nil
+}
+
+func (d *Decoder) StringSlice() ([]string, error) {
+	size, err := d.Int16()
+	if err != nil {
+		return nil, err
+	}
+	a := make([]string, size)
+	for i := range a {
+		v, err := d.String()
+		if err != nil {
+			return nil, err
+		}
+		a[i] = v
+	}
+	return a, nil
+}
+
+func (d *Decoder) read(size int) ([]byte, error) {
+	b := make([]byte, size)
```

Comment:
Since this is a private method, you can let the caller pass in the slice (which can be allocated on the stack) instead of making a new one here.